### PR TITLE
feature: scrub logs

### PR DESCRIPTION
**Pull request by saikonen**
introduces `logs scrub run_id/step/task_id` command for scrubbing logs of tasks.

by default it deletes both stdout/stderr for the latest attempt of a task.

Final feature set:
- `logs [pathspec]` works as previously, with the new default subcommand custom click group
- `logs show [pathspec]` is the default subcommand for logs cmd group
- `logs scrub [pathspec]` will overwrite logs with a redacted message

also adds support for both show/scrub commands to specify `--attempt [number]` to target a specific attempt of a task, instead of the latest.

### CODE REVIEW

**Comment by savingoyal** [comment]
@saikonen we could introduce a top-level `delete` command that has a sub command - `logs`? wdyt? also, the default behavior could be for `--latest-attempt` and users can pass in `--all-attempts` to wipe out logs for all attempts?

**Comment by saikonen** [comment]
this now introduces `logs` as a command group, with a default fallback to the new `logs show` command in case no subcommmand is specified.

the following command still works the same
```
python LongLog.py logs 1713532810109387/start/1
python LongLog.py logs 1713532810109387/start/1 --stderr
```
but now we can also support the following

```
python LongLog.py logs delete 1713532810109387/start/1
```

File: metaflow/plugins/logs_cli.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+    help="Scrub stdout/stderr produced by a task or all tasks in a step. "
+    "The format for input-path is either <run_id>/<step_name> or "
+    "<run_id>/<step_name>/<task_id>."
+)
+@click.argument("input-path")
+@click.option(
+    "--stdout/--no-stdout",
+    default=False,
+    show_default=True,
+    help="Show stdout of the task.",
```
Comment (savingoyal) [inline metaflow/plugins/logs_cli.py:220]: scrub instead of show? (here and below?)

File: metaflow/plugins/logs_cli.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+    # Pathspec can either be run_id/step_name or run_id/step_name/task_id.
+    parts = input_path.split("/")
+    if len(parts) == 2:
+        run_id, step_name = parts
+        task_id = None
+    elif len(parts) == 3:
+        run_id, step_name, task_id = parts
+    else:
+        raise CommandException(
+            "input_path should either be run_id/step_name "
```
Comment (savingoyal) [inline metaflow/plugins/logs_cli.py:257]: `input-path`?

File: metaflow/plugins/logs_cli.py [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+        ds_list = list(datastore_set)
+
+    if ds_list:
+        for ds in ds_list:
+            for stream in streams:
+                ds.scrub_logs(LOG_SOURCES, stream)
+        echo("Logs have been scrubbed.")
+
+    else:
+        raise CommandException(
```
Comment (savingoyal) [inline metaflow/plugins/logs_cli.py:286]: what is the behavior for tasks that haven't finished yet? a few different cases here - 
1. the attempt is still in flight
2. the attempt has finished but we expect further attempts

File: metaflow/datastore/task_datastore.py [hunk 4]
```diff
@@ -750,6 +762,33 @@ def save_logs(self, logsource, stream_data):
...
+
+        # Legacy log paths
+        legacy_log = self._metadata_name_for_attempt(
+            "%s.log" % stream, attempt_override
+        )
+        path_logsources[legacy_log] = stream
+
+        paths = path_logsources.keys()
+
+        deleted_paths = self._delete_file(paths, add_attempt=False)
```
Comment (savingoyal) [inline metaflow/datastore/task_datastore.py:783]: do we need to delete? can't we just overwrite?

**Review by savingoyal — COMMENTED** [review]
added a few comments

File: metaflow/plugins/logs_cli.py (hunk 2 above)
Comment (saikonen) [reply metaflow/plugins/logs_cli.py:273]: same as in the existing 'show' command. should this be reworded to something else? pathspec for example?

File: metaflow/plugins/logs_cli.py (hunk 1 above)
Comment (saikonen) [reply metaflow/plugins/logs_cli.py]: fixed.

File: metaflow/plugins/logs_cli.py (hunk 3 above)
Comment (saikonen) [reply metaflow/plugins/logs_cli.py:300]: one approach would be that we instead init the datastore with `allow_not_done=False` ?

File: metaflow/datastore/task_datastore.py (hunk 4 above)
Comment (saikonen) [reply metaflow/datastore/task_datastore.py]: turns out yes, we can simply overwrite. I cleaned up all the delete related additions to the datastore now, instead relying on `save_bytes(..., allow_overwrite=True)` to do the heavy lifting

File: metaflow/plugins/logs_cli.py (hunk 3 above)
Comment (saikonen) [reply metaflow/plugins/logs_cli.py:300]: bunch of fixes regarding this. How things work now:
we rely on `allow_not_done=False` and check when initializing the datastore in `mode="d"` that the latest / specified attempt is done, otherwise we raise an error.

- trying to `logs scrub [pathspec]` on a running task will lead to an error.
- `logs scrub [pathspec] --attempt x` will work even on a running task, as long as the specified attempt has finished.

Comment (saikonen) [reply metaflow/plugins/logs_cli.py:300]: Note: I also tried simply allowing the scrubbing of logs for running tasks, but due to how logs are written in bursts from a buffer, the effects of overwriting a log with the `[REDACTED]` message were being nullified, as every log write cycle would simply rewrite the scrubbed log lines if they were still in buffer.

File: metaflow/datastore/flow_datastore.py [hunk 5]
```diff
@@ -93,6 +93,8 @@ def get_latest_task_datastores(
         allow_not_done : bool, optional
             If True, returns the latest attempt of a task even if that attempt
             wasn't marked as done, by default False
+        mode : str, optional
```
Comment (romain-intel) [inline metaflow/datastore/flow_datastore.py:96]: nit: this is incorrect in terms of type hint, it should be: "mode : str, default "r"" or something like that.

File: metaflow/datastore/flow_datastore.py [hunk 6]
```diff
@@ -93,6 +93,8 @@ def get_latest_task_datastores(
...
+            Mode to initialize the returned TaskDataStores in. Default "r"
```
Comment (romain-intel) [inline metaflow/datastore/flow_datastore.py:97]: can remove default here.

File: metaflow/datastore/task_datastore.py [hunk 7]
```diff
@@ -173,6 +173,26 @@ def __init__(
...
+                for i in range(metaflow_config.MAX_ATTEMPTS):
+                    check_meta = self._metadata_name_for_attempt(
+                        self.METADATA_ATTEMPT_SUFFIX, i
+                    )
+                    if self.has_metadata(check_meta, add_attempt=False):
+                        self._attempt = i
+
+            # Do not allow destructive operations on the datastore if attempt is still in flight
+            # and we explicitly did not allow operating on running tasks.
+            if not allow_not_done and not self.has_metadata(self.METADATA_DONE_SUFFIX):
```
Comment (romain-intel) [inline metaflow/datastore/task_datastore.py:190]: is there a case where we want to delete logs for running tasks? It seems in delete mode we may want to ingore `allow_not_done`?

File: metaflow/plugins/logs_cli.py [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+
+from ..mflog import mflog, LOG_SOURCES
+
+# main motivation from https://github.com/pallets/click/issues/430
+# in order to support a default command being called for a Click group.
+#
+# NOTE: We need this in order to not introduce breaking changes to existing CLI, as we wanted to
+# nest both existing `logs` and the new `logs scrub` under a shared group, but `logs` already has
+# a well defined behavior of showing the logs.
+class CustomGroup(click.Group):
```
Comment (romain-intel) [inline metaflow/plugins/logs_cli.py:16]: we need to check if this is handled by @madhur-ob 's click parsing thing as well (should be but mentioning).

File: metaflow/plugins/logs_cli.py [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+    help="Scrub stdout/stderr produced by a task or all tasks in a step. "
+    "The format for input-path is either <run_id>/<step_name> or "
+    "<run_id>/<step_name>/<task_id>."
+)
+@click.argument("input-path")
+@click.option(
+    "--stdout/--no-stdout",
+    default=False,
+    show_default=True,
+    help="Scrub stdout of the task.",
```
Comment (romain-intel) [inline metaflow/plugins/logs_cli.py:229]: nit: step or task

File: metaflow/plugins/logs_cli.py [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+    help="Show stdout of the task.",
+)
+@click.option(
+    "--stderr/--no-stderr",
+    default=False,
+    show_default=True,
+    help="Show stderr of the task.",
+)
+@click.option(
+    "--both/--no-both",
```
Comment (romain-intel) [inline metaflow/plugins/logs_cli.py:109]: nit: I would add a comment (and in the next one) that this is overridden by specifying anything wrt stdout/stderr (I know it wasn't there before but unclear on the priority). Not suggesting changing the API but clarifying behavior.

File: metaflow/plugins/logs_cli.py (hunk 3 above)
Comment (romain-intel) [reply metaflow/plugins/logs_cli.py:300]: I would definiltely not advocate for allowing the redaction of in-flight logs. THere are just too many complications. We would have to prevent writing to a log that has been redacted which we can't do with past versions of metaflow so there would be this big gapping hole.

File: metaflow/plugins/logs_cli.py [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+    if ds_list:
+        for ds in ds_list:
+            for stream in streams:
+                ds.scrub_logs(LOG_SOURCES, stream)
+        echo("Logs have been scrubbed.")
+
+    else:
+        raise CommandException(
+            "No Tasks found at the given path -- "
+            "either none exist or none have started yet"
```
Comment (romain-intel) [inline metaflow/plugins/logs_cli.py:302]: or they haven't finished?

File: metaflow/plugins/logs_cli.py [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+        )
+
+
+# TODO - move step and init under a separate 'internal' subcommand
```
Comment (romain-intel) [inline metaflow/plugins/logs_cli.py:306]: ?

File: metaflow/plugins/logs_cli.py (hunk 3 above)
Comment (romain-intel) [reply metaflow/plugins/logs_cli.py:300]: I like the current behavior described @saikonen -- ie: can scrub past completed attempts with `--attempt`; can scrub latest attempt if finished with no arg and can't scrub any other case.

File: metaflow/plugins/logs_cli.py [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+        ]
+    else:
+        ds_list = obj.flow_datastore.get_latest_task_datastores(
+            run_id=run_id, steps=[step_name], attempt=attempt, mode="d"
+        )
+
+    if ds_list:
+        for ds in ds_list:
+            for stream in streams:
+                ds.scrub_logs(LOG_SOURCES, stream)
```
Comment (romain-intel) [inline metaflow/plugins/logs_cli.py:296]: This is a bit annoying in the case some logs get scrubbed but not others. I would do this in a try-catch and do as much as possible but report partial failures (maybe with echo_always?)

File: metaflow/plugins/logs_cli.py [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+    if stdout:
+        types.add("stdout")
+        both = False
+    if stderr:
+        types.add("stderr")
+        both = False
+    if both:
+        types.update(("stdout", "stderr"))
+
+    streams = list(sorted(types, reverse=True))
```
Comment (romain-intel) [inline metaflow/plugins/logs_cli.py:262]: what is the rationale behind sorting?

**Review by romain-intel — COMMENTED** [review]
No issues with the larger feature -- a few comments here and there.

File: metaflow/datastore/task_datastore.py (hunk 7 above)
Comment (saikonen) [reply metaflow/datastore/task_datastore.py:190]: I left this in mainly in order to keep the datastore side more generic (allowing delete for in-flight attempts as well, though no use-case ). It is definitely not required for this feature, so I could remove it

File: metaflow/plugins/logs_cli.py (hunk 12 above)
Comment (saikonen) [reply metaflow/plugins/logs_cli.py:306]: leftovers from `cli.py`. Cleaned up.

**Review by savingoyal — DISMISSED** [review]

File: metaflow/plugins/logs_cli.py [hunk 15]
```diff
@@ -0,0 +1,356 @@
...
+    show_default=False,
+    help="Attempt number of a task to scrub, defaults to the latest attempt.",
+)
+@click.option(
+    "--latest/--all",
+    default=True,
+    show_default=False,
+    help="Scrub latest/all attempts of a step or task",
+)
+@click.option(
```
Comment (savingoyal) [inline metaflow/plugins/logs_cli.py:256]: will this also try to scrub in-flight tasks?

File: metaflow/plugins/logs_cli.py (hunk 8 above)
Comment (saikonen) [reply metaflow/plugins/logs_cli.py:16]: tested, and this does not work as-is with the low level click api due to the fallback logic happening when invoking the command. Calling the explicit subcommand works as expected though, but the group level fallback fails

```
from metaflow.metaflow_runner import MetaflowAPI
from metaflow.cli import start

if __name__ == "__main__":

    api = MetaflowAPI.from_cli("AttemptLog.py", start)

    # works
    cmd = api().logs().show(input_path="217190/start")
    print(cmd)

    # does not
    cmd = api().logs(input_path="217190/start")
    print(cmd)
```

Will look into how easy of a fix there could be to support it

File: metaflow/datastore/flow_datastore.py [hunk 16]
```diff
@@ -93,6 +100,12 @@ def get_latest_task_datastores(
         allow_not_done : bool, optional
             If True, returns the latest attempt of a task even if that attempt
             wasn't marked as done, by default False
+        attempt : int, optional
+            Attempt number of the tasks to return. Overrides returning the latest TaskDataStore.
```
Comment (romain-intel) [inline metaflow/datastore/flow_datastore.py:104]: I would rephrase as, if not provided, returns latest attempt

File: metaflow/datastore/flow_datastore.py [hunk 17]
```diff
@@ -93,6 +100,12 @@ def get_latest_task_datastores(
...
+        include_prior : boolean, default False
+            Flag to determine if previous attempts TaskDataStores should be returned along the latest one
```
Comment (romain-intel) [inline metaflow/datastore/flow_datastore.py:106]: rephrase as: If True, returns all attempts up to and including attempt.

File: metaflow/datastore/flow_datastore.py [hunk 18]
```diff
@@ -68,7 +68,14 @@ def datastore_root(self):
         return self._storage_impl.datastore_root
 
     def get_latest_task_datastores(
```
Comment (romain-intel) [inline metaflow/datastore/flow_datastore.py:70]: rename as `get_task_datastores`

File: metaflow/datastore/flow_datastore.py [hunk 19]
```diff
@@ -127,7 +140,12 @@ def get_latest_task_datastores(
             ]
         urls = []
         for task_url in task_urls:
-            for attempt in range(metaflow_config.MAX_ATTEMPTS):
+            # parse content urls for specific attempt only, or for all attempts in max range
+            attempt_range = range(metaflow_config.MAX_ATTEMPTS)
+            if attempt is not None:
+                attempt_range = range(attempt + 1) if include_prior else [attempt]
+
```
Comment (romain-intel) [inline metaflow/datastore/flow_datastore.py:147]: I would add a check to see if attempt is less than or equal to MAX_ATTEMPTS - 1

**Review by romain-intel — COMMENTED** [review]
I like this refactor. I think with a couple name changes, it makes a lot more sense and the code looks clear imho.

File: metaflow/plugins/logs_cli.py (hunk 8 above)
Comment (romain-intel) [reply metaflow/plugins/logs_cli.py:16]: Cool -- we don't need to fix it for this but ya, we need to probbaly have it working at some point. As in, the low level api should be able to handle all our cases.

File: metaflow/datastore/flow_datastore.py (hunk 18 above)
Comment (saikonen) [reply metaflow/datastore/flow_datastore.py:70]: should we also keep `get_latest_task_datastores` which can now just be a call to `get_task_datastores(..., attempt=None, include_prior=False)`

or is `get_latest_task_datastores` not considered public and we should be fine to ditch it? OSS codebase only has two references to this so the internal rename is simple enough. Slightly concerned if something external is relying on this though.

Comment (saikonen) [reply metaflow/datastore/flow_datastore.py]: renamed based on discussions.

**Review by savingoyal — APPROVED** [review]
